### DISCUSSION TITLE
4-channel AAC streaming issue

### DISCUSSION THREAD

**Discussion by andriesess** [pschatzmann/arduino-audio-tools#1578] [Q&A] [answered]
I'm trying to stream 4-channel AAC audio (16-bit, 48kHz) over WiFi to two I2S ports by trying to combine two of your examples:

[streams-sd_wav4-i2s](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-sd_wav4-i2s)
[streams-url_aac-i2s](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-communication/http-client/streams-url_aac-i2s)

Sadly my ESP32 S3 board does not have an SD card slot so I can't truly test the 4-channel wav, however the goal is to stream 4-channel compressed audio (like AAC) over WiFi.

I tried multiple variations, I have to say I'm not too experienced, but for some reason when I send a stream with more than 2 channels it fails to work with AAC (or any other codec I tried, like OGG). The copier is reporting normal copies, but we're not hearing any audio, nor does the oscilloscope show any signals on WS, BCK or DATA. Is this a limitation of the AACHelix codec?

Is it überhaupt possible to use a single 4-channel AAC stream, but split it up for two I2S ports?
If it is not possible with AACHelix, do you have any suggestions or alternatives besides WAV to achieve this goal?

I do have access to an oscilloscope, which was quite helpful to inspect the emitted samplerates, bitdepths, etc.

in [this comment](https://github.com/pschatzmann/arduino-audio-tools/discussions/330#discussioncomment-3536197) you suggest converting the 4-channel WAV example to AAC which we try as follows....

**Device Description**
QTPy ESP32-S3 with PS RAM
4x MAX98357A DAC's

**This is the code:**

#include "AudioTools.h"
#include "AudioCodecs/CodecAACHelix.h"
#include "driver/gpio.h"

URLStream url("myWifi","myPassword");

AudioInfo info4(48000, 4, 16);
AudioInfo info2(48000, 2, 16);

I2SStream i2s_1;
I2SStream i2s_2;

ChannelsSelectOutput out;

EncodedAudioOutput decoder(&out, new AACDecoderHelix());  // Decoding stream

StreamCopy copier; // copy url to decoder

void setup(){
  Serial.begin(115200);

  AudioLogger::instance().begin(Serial, AudioLogger::Info);
  Serial.println("setup started...");

  // setup i2s
  auto config1 = i2s_1.defaultConfig(TX_MODE);
  config1.copyFrom(info2);
  config1.pin_ws = 9;
  config1.pin_bck = 17;
  config1.pin_data = 18;
  config1.port_no = 0;
  //config.mode = I2S_STD_FORMAT;
  i2s_1.begin(config1);

  auto config2 = i2s_2.defaultConfig(TX_MODE);
  config2.copyFrom(info2);
  config2.pin_ws = 36;
  config2.pin_bck = 37;
  config2.pin_data = 35;
  config2.port_no = 1;
  //config.mode = I2S_STD_FORMAT;
  i2s_2.begin(config2);

  out.addOutput(i2s_1, 0, 1);
  out.addOutput(i2s_2, 2, 3);
 
  //out.addNotifyAudioChange(i2s_1); 
  //out.addNotifyAudioChange(i2s_2);
 
  out.begin(info4);
 
  decoder.begin(info4);

  url.begin("http://192.168.129.76:8000/stream","audio/aac");  
  copier.begin(decoder, url);

  Serial.println("started...");
}

void loop(){
  copier.copy();
}

**Comment by pschatzmann** [comment, marked as answer]
I would not know since I have never tried: In theory AAC should support more then 2 channels.
Are you getting any error messages ? If not  I suggest to have a look at the log to figure out what's going on. If info does not provide enough granularity you can try with debug.

You could also try to send thee output to a CsvStream to double check...
